lib/common/ipc_client.c

```diff
@@ -31,6 +31,672 @@
+
+    if ((api == NULL) || (api->dispatch_type != pcmk_ipc_dispatch_poll)) {
+        return EINVAL;
+    }
+    pollfd.fd = crm_ipc_get_fd(api->ipc);
+    pollfd.events = POLLIN;
+    rc = poll(&pollfd, 1, timeout_ms);
+    if (rc < 0) {
+        return errno;
+    } else if (rc == 0) {
+        return EAGAIN;
+    }
+    return pcmk_rc_ok;
+}
+
+/*!
+ * \brief Dispatch available messages on an IPC connection (without main loop)
+ *
+ * \param[in]  api  IPC API connection
+ *
+ * \return Standard Pacemaker return code
+ *
+ * \note Callers of pcmk_connect_ipc() using pcmk_ipc_dispatch_poll should call
+ *       this function when IPC data is available.
+ */
+void
+pcmk_dispatch_ipc(pcmk_ipc_api_t *api)
+{
+    if (api != NULL) {
+        while (crm_ipc_ready(api->ipc)) {
+            if (crm_ipc_read(api->ipc) > 0) {
+                dispatch_ipc_data(crm_ipc_buffer(api->ipc), 0, api);
+            }
+        }
+    }
+}
+
+// \return Standard Pacemaker return code
+static int
+connect_with_main_loop(pcmk_ipc_api_t *api)
+{
+    int rc;
+
+    struct ipc_client_callbacks callbacks = {
+        .dispatch = dispatch_ipc_data,
+        .destroy = ipc_post_disconnect,
+    };
+
+    rc = pcmk__add_mainloop_ipc(api->ipc, G_PRIORITY_DEFAULT, api,
+                                &callbacks, &(api->mainloop_io));
+    if (rc != pcmk_rc_ok) {
+        return rc;
+    }
+    crm_debug("Connected to %s IPC (attached to main loop)",
+              pcmk_ipc_name(api, true));
+    /* After this point, api->mainloop_io owns api->ipc, so api->ipc
+     * should not be explicitly freed.
+     */
+    return pcmk_rc_ok;
+}
+
+// \return Standard Pacemaker return code
+static int
+connect_without_main_loop(pcmk_ipc_api_t *api)
+{
+    int rc;
+
+    if (!crm_ipc_connect(api->ipc)) {
+        rc = errno;
+        crm_ipc_close(api->ipc);
+        return rc;
+    }
+    crm_debug("Connected to %s IPC (without main loop)",
+              pcmk_ipc_name(api, true));
+    return pcmk_rc_ok;
+}
+
+/*!
+ * \brief Connect to a Pacemaker daemon via IPC
+ *
+ * \param[in]  api            IPC API instance
+ * \param[out] dispatch_type  How IPC replies should be dispatched
+ *
+ * \return Standard Pacemaker return code
+ */
+int
+pcmk_connect_ipc(pcmk_ipc_api_t *api, enum pcmk_ipc_dispatch dispatch_type)
+{
+    int rc = pcmk_rc_ok;
+
+    if ((api == NULL) || (api->ipc == NULL)) {
+        crm_err("Cannot connect to uninitialized API object");
+        return EINVAL;
+    }
+
+    if (crm_ipc_connected(api->ipc)) {
+        crm_trace("Already connected to %s IPC API", pcmk_ipc_name(api, true));
+        return pcmk_rc_ok;
+    }
+
+    api->dispatch_type = dispatch_type;
+    switch (dispatch_type) {
+        case pcmk_ipc_dispatch_main:
+            rc = connect_with_main_loop(api);
+            break;
+
+        case pcmk_ipc_dispatch_sync:
+        case pcmk_ipc_dispatch_poll:
+            rc = connect_without_main_loop(api);
+            break;
+    }
```

Comment:
Do we build with whatever flags would be necessary to warn if a member of an enum is not covered by a switch statement?  Does gcc even have such a thing?  My point is that if such a flag doesn't exist, we should probably add a default here (and wherever else in this patch we're switching on an enum) to prevent weird behavior should the enums ever gain new members.

Reply:
Yes it's checked, by coverity IIRC. And some analyzer will complain about the default never being reached if all enum values have their own case. The combination of those two makes defensive programming difficult. :-/